Generate devcontainer files by default

andrewn617 opened the pull request:
https://github.com/rails/rails/issues/50566

### Motivation / Background

This Pull Request has been created to generate `devcontainer` files for new Rails apps.

### Detail

This Pull Request adds templates for devcontainer files (`devcontainer.json`, `Dockerfile` and `docker-compose.yml`) to be generated on `rails new`.

By default the devcontainer contains the following:
- A redis container for use with Kredis, Action Cable etc.
- A database (SQLite, Postgres, MySQL or MariaDB)
- A Headless chrome container for system tests
- Active Storage configured to use the local disk and with preview features working

If `rails new` is called with options to skip any of the above they will be removed from the dev container config.

The dev container can be skipped entirely with the `--skip_devcontainer` option.

### Additional information

A few changes are needed to other files generated by rails new to make sure new Rails apps can be developed both in a container and locally.

- ~Differentiate `CABLE_REDIS_URL` and `JOBS_REDIS_URL` instead of just using `REDIS_URL` for both (A PR will need to be opened to turbo-rails to make this change as well)~
- ~For MariaDB and MySQL use the `RAILS_DATABASE_HOST` if it's set (which it will be in a container) and fallback to `localhost`~
- ~For Postgres set the host (`RAILS_DATABASE_HOST`), username and password when in a container`~

Based on feedback received I reverted the above ^. Instead we will just `REDIS_URL` and `DB_HOST`.

For system tests, we need to configure the `Capybara` host and port to the selenium container. Rather than expose the implementation details of `Capybara` to the developer, we created a new `served_by` method on `SystemTestCase` to configure these.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

andrewn617 committed:
Generate a .devcontainer folder and its contents when creating a new app.

The .devcontainer folder includes everything needed to boot the app and do development in a remote container.

The container setup includes:
 - A redis container for Kredis, ActionCable etc.
 - A database (SQLite, Postgres, MySQL or MariaDB)
 - A Headless chrome container for system tests
 - Active Storage configured to use the local disk and with preview features working

If any of these options are skipped in the app setup they will not be included in the container configuration.

These files can be skipped using the `--no-devcontainer` option.

Co-authored-by: Rafael Mendonça França <rafael@franca.dev>

rafaelfranca commented on railties/lib/rails/generators/rails/app/templates/.devcontainer/devcontainer.json.tt:
```diff
@@ -0,0 +1,49 @@
+
+  "containerEnv": {
+<%- if depends_on_system_test? -%>
+    "CAPYBARA_SERVER_PORT": "45678",
```
Doesn't the trailing `,` break if there is only one item?

andrewn617 replied:
Hm, it doesn't seem to cause any problems for vscode. But I will fix it.

rafaelfranca commented on railties/lib/rails/generators/rails/app/templates/.devcontainer/devcontainer.json.tt:
```diff
@@ -0,0 +1,49 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/ruby
+{
+  "name": "Ruby",
```
This should be the application name

rafaelfranca reviewed (commented):
Is `CAPYBARA_SERVER_PORT` being used anywhere?

rafaelfranca commented:
The Capybara setting will not work because `app_host` and `app_port` isn't configured

andrewn617 commented:
> The Capybara setting will not work because `app_host` and `app_port` isn't configured

Sorry, missed that one. Will fix it.

andrewn617 committed:
Devcontainers: Introduce SystemTestCase#served_by

When working in a devcontainer, for system tests we need to manually set the host and port serving the application. Let's introduce a method for this, so we don't have to expose the implementation details of Capybara to the developer.

Co-authored-by: Rafael Mendonça França <rafael@franca.dev>

rafaelfranca commented on actionpack/lib/action_dispatch/system_test_case.rb:
```diff
@@ -159,6 +159,11 @@ def self.driven_by(driver, using: :chrome, screen_size: [1400, 1400], options: {
       self.driver = SystemTesting::Driver.new(driver, **driver_options, &capabilities)
     end
 
+    def self.served_by(host:, port:)
```
Document this

rafaelfranca commented on railties/lib/rails/generators/rails/app/templates/test/application_system_test_case.rb.tt:
```diff
@@ -1,5 +1,14 @@
+
+    driven_by :selenium, using: :headless_chrome, screen_size: [ 1400, 1400 ], options: {
+      browser: :remote,
+      url: "http://selenium:4444",
```
This should be another ENV

rafaelfranca commented on railties/lib/rails/generators/rails/db/system/change/change_generator.rb:
```diff
@@ -13,6 +13,9 @@ class ChangeGenerator < Base # :nodoc:
           class_option :to, required: true,
             desc: "The database system to switch to."
 
+          class_option :skip_devcontainer, type: :boolean, default: true,
```
```suggestion
          class_option :devcontainer, type: :boolean, default: false,
```

This would give us `--no-devcontainer`. We should do the same on the app generator as well

andrewn617 replied:
TIL :D

rafaelfranca commented on railties/lib/rails/generators/app_base.rb:
```diff
@@ -109,6 +109,9 @@ def self.add_shared_options_for(name)
                                            desc: "Skip GitHub CI files"
 
+        class_option :devcontainer,        type: :boolean, default: true,
+                                           desc: "Generator devcontainer files"
```
```suggestion
                                           desc: "Generate devcontainer files"
```

rafaelfranca commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -1229,6 +1232,145 @@ def test_name_option
+  end
+
+  def test_devcontainer_no_redis_skipping_action_cable_and_active_job
+    run_generator [ destination_root, "--skip_action_cable", "--skip_active_job" ]
```
```suggestion
    run_generator [ destination_root, "--skip-action-cable", "--skip-active-job" ]
```

rafaelfranca commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -1229,6 +1232,145 @@ def test_name_option
+  end
+
+  def test_devcontainer_no_jobs_redis_url_when_skipping_active_job
+    run_generator [ destination_root, "--skip_active_job" ]
```
```suggestion
    run_generator [ destination_root, "--skip-active-job" ]
```

rafaelfranca commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -1229,6 +1232,145 @@ def test_name_option
+  end
+
+  def test_devcontainer_no_cable_redis_url_when_skipping_action_cable
+    run_generator [ destination_root, "--skip_action_cable" ]
```
```suggestion
    run_generator [ destination_root, "--skip-action-cable" ]
```

rafaelfranca commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -1229,6 +1232,145 @@ def test_name_option
+  end
+
+  def test_devcontainer_no_Dockerfile_packages_when_skipping_active_storage
+    run_generator [ destination_root, "--skip_active_storage" ]
```
```suggestion
    run_generator [ destination_root, "--skip-active-storage" ]
```

gabrielso commented on railties/lib/rails/generators/rails/app/templates/.devcontainer/devcontainer.json.tt:
```diff
@@ -0,0 +1,34 @@
+  // "forwardPorts": [],
+
+  // Configure tool-specific properties.
+  // "customizations": {},
```
If the target is to support Dev Containers in VS Code, then it could be valuable to add the Ruby LSP extension by default...

```json
"customizations": {
  "vscode": {
    "extensions": [
      "Shopify.ruby-lsp"
    ]
  }
}
```

This way, default debug configurations can also be added in `.vscode/launch.json` to debug the app and tests. I may have a working example of this setup in a private repo I could share, if there's interest.

rafaelfranca replied:
It is already in the ruby feature present in the base image.

rafaelfranca replied:
We decided to not add the launch.json yet because it isn't working in a way that is satisfactory to me. When I get my team to fix it, we will add all LSP features. Also see #50453

viktorianer commented on railties/lib/rails/generators/app_base.rb:
```diff
@@ -109,6 +109,9 @@ def self.add_shared_options_for(name)
         class_option :skip_ci,             type: :boolean, default: nil,
                                            desc: "Skip GitHub CI files"
 
+        class_option :devcontainer,        type: :boolean, default: true,
```
Should we not follow the naming conventions here?

> This would give us `--no-devcontainer`. We should do the same on the app generator as well

If `--no-` is the default, all others should follow this naming too?

viktorianer commented on railties/lib/rails/generators/rails/app/templates/.devcontainer/docker-compose.yml.tt:
```diff
@@ -0,0 +1,78 @@
+  redis:
+    image: redis:7.2
+    restart: unless-stopped
+    networks:
+      - default
+    ports:
+      - 6379:6379
+<%- end -%>
+
+<%- if options.database == "postgresql" -%>
+  postgres:
+    image: postgres:16.1
+    restart: unless-stopped
+    networks:
+      - default
+    environment:
+      POSTGRES_USER: postgres
+      POSTGRES_PASSWORD: postgres
+<%- end -%>
+
+<%- if options.database == "mysql" -%>
+  mysql:
+    image: mysql/mysql-server:8.0
+    restart: unless-stopped
+    environment:
+      MYSQL_ALLOW_EMPTY_PASSWORD: true
+      MYSQL_ROOT_HOST: "%"
+    networks:
+      - default
+<%- end -%>
+
+<%- if options.database == "trilogy" -%>
+  mariadb:
+    image: mariadb:10.5
+    restart: unless-stopped
+    networks:
+      - default
+    environment:
+      MARIADB_ALLOW_EMPTY_ROOT_PASSWORD: true
+<%- end -%>
```
Moving all services in `devcontainer_dependencies` (`depends_on`) to a `docker-compose.yml.tt` root folder file, would improve dev environment and adoption to docker.

Because, having all service dependencies in root `docker-compose.yml.tt` allow us to start the services only with `docker compose` and the Rails server locally.  Meaning, along with dev containers, we could use service with `docker compose up -d`, so every developer could choose between:

- local server and local services
- dev containers in Docker
- local server and services in Docker

without changing any line in settings.

andrewn617 replied:
The goal of this work is to create a light weight development environment that is easy to get working without installing many dependencies, especially for people new to Rails and even new to coding. While I can definitely see the value of having the Rails app running locally and other services in the container, it is a more advance setup and not exactly with the scope of our goals for this work. So I am not sure we want to make that a default part of the configuration and thus something the Rails team has to support going forward.

Rafael is away for a couple days, let's get his opinion when he is back.

viktorianer replied:
As you can see in RubyGems repo, there are no more dependencies, and it is actually more `light weight development environment` as going with Dev Containers. In case of RubyGems, I was working on a feature there, I was not able to run Dev Containers, but I could easily run `docker compose up` and working on this project.

Would love to hear Rafael's opinion too.

palkan replied:
You can still use `docker compose up -f .devcontainer/docker-compose.yml -d <service>`.

Having dev env configuration in a dedicated folder (like `.devcontainer`) would prevent from potential conflicts with production `docker-compose.yml` (which is usually kept in the root) or other Docker configurations.

viktorianer replied:
Yes, this sounds good to me too.

viktorianer commented on railties/lib/rails/generators/rails/app/templates/config/cable.yml.tt:
```diff
@@ -6,5 +6,5 @@ test:
 production:
   adapter: redis
-  url: <%%= ENV.fetch("REDIS_URL") { "redis://localhost:6379/1" } %>
+  url: <%%= ENV.fetch("CABLE_REDIS_URL") { "redis://localhost:6379/1" } %>
```
That will potentially break all running apps in production, without any notice? Because, `REDIS_URL` is so common used.

andrewn617 replied:
We are going to just still with `REDIS_URL`. However, this is just a template for generating new apps / installing action cable, so changes here should not impact existing apps already using action cable.

viktorianer replied:
You are right, I was thinking about upgrades. But anyway, like the fact, you are reverted it to `REDIS_URL` 👍🏼 .

viktorianer commented on railties/lib/rails/generators/rails/app/templates/config/databases/postgresql.yml.tt:
```diff
@@ -18,6 +18,13 @@ default: &default
   # For details on connection pooling, see Rails configuration guide
   # https://guides.rubyonrails.org/configuring.html#database-pooling
   pool: <%%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
+<% if options[:devcontainer] -%>
+  <%% if ENV["RAILS_DATABASE_HOST"] %>
+  host: <%%= ENV["RAILS_DATABASE_HOST"] %>
+  username: postgres
+  password: postgres
+  <%% end %>
+<% end %>
```
I think, this should be used in `development` and `test` environment only.

andrewn617 replied:
I think we can expect that the `DB_HOST` (I changed it from `RAILS_DATABASE_HOST` based on some feedback) env variable will not be set in production.

palkan commented on railties/lib/rails/generators/rails/app/templates/.devcontainer/docker-compose.yml.tt:
```diff
@@ -0,0 +1,78 @@
+    networks:
+      - default
+    ports:
+      - 6379:6379
```
Why do we expose Redis port to the host?

viktorianer replied:
See the comment above:

> # Use "forwardPorts" in **devcontainer.json** to forward an app port locally.
> # (Adding the "ports" property to this file will not forward from a Codespace.)

palkan replied:
> to forward an app port locally

Yeah, but why accessing internal (Docker) Redis locally?

I'm worried because that's a common situation when developers always have Redis running on their machine on the default port, so, trying to expose a Redis one on the same port would result in a conflict.

andrewn617 replied:
Yes, I think you're right it is not needed. I am going to remove it.

viktorianer replied:
> See the comment above:
> 
> > ```shell
> > # Use "forwardPorts" in **devcontainer.json** to forward an app port locally.
> > # (Adding the "ports" property to this file will not forward from a Codespace.)
> > ```
> 
> But, in order to prevent third party access, I would suggest to bind all ports to `127.0.0.1`, e.g.,:
> 
> `"127.0.0.1:6379:6379"`

### Compose

Just in case, you are going to consider https://github.com/rails/rails/pull/50914#discussion_r1479818590.

I'm just realizing, when using Dev Containers, the ports are forward to host, even `forwardPorts` is not present. So, without binding to `"127.0.0.1:5432:5432"`, the database would be exposed to everyone in the network. 🥇 

We should avoid this!

### Dev Container CLI

Also, when there is no definition for port forwarding in compose file, e.g., `ports: 3000:3000` or better `"127.0.0.1:3000:3000"`, working with Dev Container CLI is not possible. The port will be not forward, even you define `forwardPorts` in order to access the Rails server. See https://github.com/devcontainers/cli/issues/186.

### Conclusion

Please, add port forwarding to `compose.yaml` file, not to `devcontainer.json`. 

By prepending it with `127.0.0.1:`, we can even secure expose database to host, in order to support tools like pgAdmin or other.

palkan commented on railties/lib/rails/generators/rails/app/app_generator.rb:
```diff
@@ -592,6 +605,35 @@ def after_bundle(&block) # :doc:
+        @devcontainer_variables["CAPYBARA_SERVER_PORT"] = "45678" if depends_on_system_test?
+        @devcontainer_variables["SELENIUM_HOST"] = "selenium" if depends_on_system_test?
+        @devcontainer_variables["REDIS_URL"] = "redis://redis:6379/1" unless options.skip_active_job? && options.skip_action_cable?
+        @devcontainer_variables["DB_HOST"] = "postgres" if options.database == "postgresql"
```
Why don't we use `DATABASE_URL`?

andrewn617 replied:
Trilogy and mysql2 templates use `host` already, so we will continue with that and use `host` in the postgres template to remain consistent.

palkan commented on railties/lib/rails/generators/rails/app/templates/.devcontainer/Dockerfile.tt:
```diff
@@ -0,0 +1,14 @@
+FROM mcr.microsoft.com/devcontainers/ruby:1-3-bookworm
```
Shouldn't it be "3-1"?

```suggestion
FROM mcr.microsoft.com/devcontainers/ruby:3-1-bookworm
```

And, I guess, we need to use the current app's Ruby version here.

andrewn617 replied:
No. The 1 here indicates the version of the image, not the version of ruby.

We plan to host our own images prior to this feature's release, so this is really just a placeholder for now.

palkan commented on actionpack/lib/action_dispatch/system_test_case.rb:
```diff
@@ -159,6 +159,14 @@ def self.driven_by(driver, using: :chrome, screen_size: [1400, 1400], options: {
+    # By default this is localhost. This method allows the host and port to be specified manually.
+    def self.served_by(host:, port:)
+      Capybara.server_host = host
+      Capybara.server_port = port
```
Shouldn't we also configure `Capybara.app_host` (so the Selenium container would be able to access the Rails test server)? 

See also https://evilmartians.com/chronicles/system-of-a-test-setting-up-end-to-end-rails-testing#dockerizing-system-tests

andrewn617 replied:
Is it needed? The tests run with this configuration.

palkan replied:
> The tests run with this configuration.

yeah, that's probably due to the use of the shared network and the exposed test server port (so, `localhost:45678` works from the chrome service).

palkan commented on railties/lib/rails/generators/rails/app/templates/.devcontainer/docker-compose.yml.tt:
```diff
@@ -0,0 +1,78 @@
+services:
```
What about naming this file `compose.yaml` (it has been [the default name](https://docs.docker.com/compose/compose-application-model/#the-compose-file) for Compose configuration for a while).

viktorianer replied:
And, we need data `volumes:` for the database?!

viktorianer commented on railties/lib/rails/generators/rails/app/templates/.devcontainer/devcontainer.json.tt:
```diff
@@ -0,0 +1,34 @@
+    "ghcr.io/devcontainers/features/github-cli:1": {}
+  },
+
+<%- if !devcontainer_variables.empty? -%>
+  "containerEnv": {
+    <%= devcontainer_variables.map { |key, value| "\"#{key}\": \"#{value}\"" }.join(",\n    ") %>
+  },
+<%- end -%>
```
Adding it to compose file directly, e.g., as an extension, would make it possible to use the same compose file without Dev Containers.

```yml
x-env: &app-env
  DB_HOST: postgres
  REDIS_URL: redis://redis:6379/1

services:
      rails-app:
        environment:
          <<: *app-env

```

I know, it is not in the scope of this PR to support `docker compose -f .devcontainer/compose.yaml up -d`, but this small difference would make it possible.

viktorianer commented on railties/lib/rails/generators/database.rb:
```diff
@@ -90,6 +90,14 @@ def mysql_socket
           ].find { |f| File.exist?(f) } unless Gem.win_platform?
         end
+
+        def mysql_database_host
+          if options[:devcontainer]
+            "<%= ENV.fetch(\"DB_HOST\") { \"localhost\" } %>"
+          else
+            "localhost"
+          end
+        end
```
Why not using it for all database configs?

```ruby
def database_host
 "<%= ENV.fetch(\"DB_HOST\") { \"localhost\" } %>"
end
```

Would work for `templates/config/databases/postgresql.yml.tt` as well as for all other templates.

But, I am trying to address this issue differently in https://github.com/rails/rails/pull/50975. Have a look.

andrewn617 replied:
Unfortunately it will not work - for local postgresql on MacOs the connection is by socket using the macos user to authenticate - so no host option is supplied (look at how the `postgres.yml.tt` looks on main to see what I mean). We want need to preserve that behaviour in case the dev is not using a dev container, which is why you see the `if ENV["DB_HOST"]` in the postgres template wrapping the host, user and password fields.

andrewn617 committed:
Rename .devcontainer/docker_compose.yml to compose.yaml

compose.yaml is considered the preferred name for the Compose file.

See https://docs.docker.com/compose/compose-application-model/#the-compose-file for details.

rafaelfranca commented on railties/lib/rails/generators/rails/app/templates/.devcontainer/compose.yaml.tt:
```diff
@@ -0,0 +1,56 @@
+    # Use "forwardPorts" in **devcontainer.json** to forward an app port locally.
+    # (Adding the "ports" property to this file will not forward from a Codespace.)
+    ports:
+    - 3000:3000
```
Do we need the 3000 port? VScode maps automatically as soon as you run `rails s`

rafaelfranca committed:
Temporarily remove `ruby` from Gemfile

See 707b5cb60616c86cd554ada2ef8754dd0ab5e72a for more details.

mhenrixon commented:
Curious question: I couldn't see a way to add devcontainers to an existing project, is that by design?

rafaelfranca commented:
> Curious question: I couldn't see a way to add devcontainers to an existing project, is that by design?

At the moment, yes

rafaelfranca commented on railties/test/fixtures/.devcontainer/compose.yaml:
```diff
@@ -0,0 +1,45 @@
+    # Use "forwardPorts" in **devcontainer.json** to forward an app port locally.
+    # (Adding the "ports" property to this file will not forward from a Codespace.)
+    ports:
+    - 45678:45678
```
If you want that, feel free to change in your app. Rails by design binds the development environment to 0.0.0.0.

rafaelfranca replied:
Although this might not be necessary. We are exposing the port so the app container talks with the selenium container. If we use the bridge mode that should not be necessary. I'll investigate.

rafaelfranca replied:
Will be fixed in https://github.com/rails/rails/pull/52020. Thank you for pointing it.

rafaelfranca commented:
It is not temporary anymore. We decided to not generate ruby version in the gemfile

voxik commented:
Work for me. Thanks for clarification 👍
